File: cpp/cpp_100-999/713.cc
// C++
// leetcode 713
// https://leetcode-cn.com/problems/subarray-product-less-than-k/

class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
    	int cur = 1, n = nums.size();
    	int res = 0;
    	for(int i = 0, j = 0; i < n; i++) {
            j = max(i, j);
    		while(j < n && cur * nums[j] < k) {
    			cur *= nums[j];
    			j++;
    		}
    		
    		res += j-i;
    		// cout<<i<<" : "<<res<<endl;
    		if(cur != 1) {
    			cur /= nums[i];
    		}
    	}
    	return res;
    }
};

File: Go/Go_100-999/713.go
// Go
// leetcode 713
// https://leetcode.cn/problems/subarray-product-less-than-k/

func numSubarrayProductLessThanK(nums []int, k int) int {
	n := len(nums)
	product := 1
	res := 0
	for l, r := 0, 0; l < n; l++ {
		if r < l {
			r = l
		}
		for r < n && product*nums[r] < k {
			product *= nums[r]
			r++
		}
		res += r - l
		if l != r {
			product /= nums[l]
		}
	}
	return res
}